---
title: "enrichment_analaysis_mutation_data"
output: html_document
date: "2025-07-12"
Objective: Assess whether specific biological pathways are under selection through enrichment analysis.
Aims:
1. Run eggnog annotation on reference amino acid fasta file 
2. Extract reference genes which carry mutations from binary data matrix -- https://github.com/katherine294/psf_evolutionary_analysis/tree/main/Figure2/snippy-to-binary.Rmd
3. Run enrichment analysis on gene sets
---


RUN EGGNOG TO ANNOTATE THE REFERENCE GENOME OF CHOICE ------------------------------
Prerequisites: 
1. Download the eggnog database  -- https://github.com/eggnogdb/eggnog-mapper/wiki/eggNOG-mapper-v2.1.5-to-v2.1.13#user-content-Installation
2. Save createmappingfiles.py in a suitable directory e.g. "path/to/scripts" - https://github.com/katherine294/psf_evolutionary_analysis/tree/main/scripts
   Change the line : emapper = pd.read_csv("W163A3B1_mutations.emapper.annotations", sep='\t', comment='#') to your prefix from eggnog annotation.


```{bash}
#!/bin/bash
#SBATCH --time 72:00:00
#SBATCH --mem-per-cpu=6750M
#SBATCH --ntasks=54
#SBATCH --mail-type ALL

set -e

# Load BlueBEAR environment
module purge; module load bluebear
module load bear-apps/2022b
module load Miniforge3/24.1.2-0
module load Biopython/1.81-foss-2022b

# Activate Conda
eval "$(${EBROOTMINIFORGE3}/bin/conda shell.bash hook)"
source "${EBROOTMINIFORGE3}/etc/profile.d/mamba.sh"

# Define your environment path
CONDA_ENV_PATH="/rds/homes/k/kgh742/psf_wgs_project/conda/eggnog"

# Activate your environment
mamba activate "${CONDA_ENV_PATH}"

# Define paths
PROJECT_BASE="/rds/homes/k/kgh742/psf_wgs_project"
OUTDIR="$PROJECT_BASE/Enrichment_analysis/eggnog_output_W163"
REF_DIR="$PROJECT_BASE/reference_sequences"
EGGNOGG_DB="$PROJECT_BASE/EGGNOGG_DB" # Directory you saved your EGGNOG database to 

# Set the data directory environment variable (optional but useful)
export EGGNOG_DATA_DIR="$EGGNOGG_DB"

mkdir -p $OUTDIR

# Run eggNOG-mapper
  emapper.py \
    -i "$REF_DIR/W163a3b1.faa" \ # use protein amino acid fasta as input
    -o W163a3b1_mutations \ # Prefix
    --itype proteins \
    -m diamond \
    --cpu 50 \
    --tax_scope Bacteria \
    --output_dir "$OUTDIR"

# CREATE MAPPING FILES --------------------

cd "$OUTDIR"

python $PROJECT_BASE/scripts/createmappingfiles.py
```


Load libraries
```{r}
library(readr)
library(dplyr)
library(stringr)
library(ggplot2)

library(KEGGREST)  # KEGG analysis
library(GO.db) # GO analysis
library(AnnotationDbi) # GO analysis
library(clusterProfiler) # KEGG and GO

options(bitmapType='cairo') # Visualising graphs within an interactive R session
```


Create gene lists of genes carrying synonymous or non-synoymous mutations
````{r}
df <- read_csv("binary_mutation_matrix_psf")

# Save list of genes carrying synonymous of non-synonymous mutations  
unique_synonymous <- df %>%
  filter(str_detect(effect, "synonymous_variant")) %>%
  select(locus, effect) %>%
  distinct(locus) %>%     # Only keep unique genes
  pull(locus)    # Save as vector not dataframe

unique_nonsynonymous <- df %>%
  filter(!is.na(locus)) %>%                   # keep only rows with locus not NA
  filter(!str_detect(effect, "synonymous_variant")) %>%  # remove synonymous variants 
   distinct(locus) %>%
   pull(locus)

### SAVE SUMMARY TABLE
# --- Define subsets ---
# Total number of mutations
total_mutations <- nrow(df)

# Synonymous
synonymous <- df %>%
  filter(str_detect(effect, "synonymous_variant"))

# Non-synonymous (excluding synonymous + keeping only genic)
nonsynonymous <- df %>%
  filter(!is.na(locus)) %>%
  filter(!str_detect(effect, "synonymous_variant"))

# Intragenic (mutations that have a locus)
intergenic <- df %>%
  filter(is.na(effect))

# --- Create summary table ---
summary_table <- tibble(
  Category = c("All mutations", "Synonymous", "Non-synonymous", "Intergenic"),
  Mutation_count = c(
    nrow(df),
    nrow(synonymous),
    nrow(nonsynonymous),
    nrow(intergenic)
  ),
  Unique_genes = c(
    n_distinct(df$locus),
    n_distinct(synonymous$locus),
    n_distinct(nonsynonymous$locus),
    n_distinct(intergenic$locus)
  )
)

write_csv(summary_table, "/rds/homes/k/kgh742/psf_wgs_project/Enrichment_analysis/eggnog_output_W163/summary_table.csv")
```

KEGG analysis
```{r}
# Load KEGG mapping file 
kegg_map <- read.table("/rds/homes/k/kgh742/psf_wgs_project/Enrichment_analysis/eggnog_output_W163/gene2kegg.txt", sep = "\t", stringsAsFactors = FALSE)
colnames(kegg_map) <- c("gene", "kegg_terms")
gene2kegg <- strsplit(kegg_map$kegg_terms, ";")
names(gene2kegg) <- kegg_map$gene

# Load background test gene list
background_genes <- names(gene2kegg)

### Run enrichment on synonymous gene set ###
kegg_enrich_S <- enricher(gene = unique_synonymous,
                universe = background_genes,
                TERM2GENE = stack(gene2kegg),
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# get KEGG IDs from enrichment
kegg_ids_S <- kegg_enrich_S@result$ID  # example: ko00010, ko00020

# fetch names/descriptions
descs_S <- sapply(kegg_ids_S, function(id) {
  res <- tryCatch(keggGet(id), error = function(e) NULL)
  if (!is.null(res)) res[[1]]$NAME else NA
})

# add to results
kegg_enrich_S@result$Description <- descs_S

# Visualise
kegg_plot_S <- dotplot(kegg_enrich_S, showCategory = 20)
kegg_plot_S

# Analysing results table
sig_results_S <- as.data.frame(kegg_enrich_S)
sig_results_S <- sig_results_S[sig_results_S$p.adjust < 0.05, ]
sig_results_S

write.csv(sig_results_S, "KEGG_signif_synonymous.csv")

### Repeat enrichment analysis on non-synonymous gene set ###
kegg_enrich_NS <- enricher(gene = unique_nonsynonymous,
                universe = background_genes,
                TERM2GENE = stack(gene2kegg),
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# get KEGG IDs from enrichment
kegg_ids_NS <- kegg_enrich_NS@result$ID  # example: ko00010, ko00020

# fetch names/descriptions
descs_NS <- sapply(kegg_ids_NS, function(id) {
  res <- tryCatch(keggGet(id), error = function(e) NULL)
  if (!is.null(res)) res[[1]]$NAME else NA
})

# add to results
kegg_enrich_NS@result$Description <- descs_NS

# Visualise
kegg_plot_NS <- dotplot(kegg_enrich_NS, showCategory = 20)
kegg_plot_NS

# Analysing results table
sig_results_NS <- as.data.frame(kegg_enrich_NS)
sig_results_NS <- sig_results_NS[sig_results_NS$p.adjust < 0.05, ]
sig_results_NS

write.csv(sig_results_NS, "KEGG_signif_non-synonymous.csv")
```

GO enrichment
```{r}
# Read gene-to-GO mapping (tab-delimited: geneID <tab> GO:term1;GO:term2...)
go_map <- read.table("/rds/homes/k/kgh742/psf_wgs_project/Enrichment_analysis/eggnog_output_W163/gene2kegg.txt", sep = "\t", stringsAsFactors = FALSE)
colnames(go_map) <- c("gene", "go_terms")

# Split the GO terms
gene2go <- strsplit(go_map$go_terms, ";")
names(gene2go) <- go_map$gene

# Load background test gene list
background_go_genes <- names(gene2go)

# Run enrichment 
ego_S <- enricher(gene = unique_synonymous,
                universe = background_go_genes,
                TERM2GENE = stack(gene2go),
                pvalueCutoff = 1,   # Increase to 1 for visualising top results if non are statistically significant
                qvalueCutoff = 1)

# Add GO term descriptions
ego_S@result$Description <- AnnotationDbi::Term(ego_S@result$ID)

# Visualise
dotplot_S <- dotplot(ego_S)


# Reun enrichment on non-synonymous genes
ego_NS <- enricher(gene = unique_nonsynonymous,
                universe = background_go_genes,
                TERM2GENE = stack(gene2go),
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# Add GO term descriptions
ego_NS@result$Description <- AnnotationDbi::Term(ego_NS@result$ID)

# Visualise
dotplot_NS <- dotplot(ego_NS)


```
